File: erlang/mixi.erl
-module(mixi).
-export([get_token/4,
	 get_my_friends/1]).

get_token(AuthorizationCode,
         ClientId,
         ClientSecret,
         RedirectUri) ->
    Response = httpc:request(
                 post,
                 {
                   "https://secure.mixi-platform.com/2/token",
                   [],
                   "application/x-www-form-urlencoded",
                   lists:append([
                                 "grant_type=authorization_code",
                                 "&client_id=", ClientId,
                                 "&client_secret=", ClientSecret,
                                 "&redirect_uri=", edoc_lib:escape_uri(RedirectUri),
                                 "&code=", AuthorizationCode
                                ])
                 },
                 [],
                 []),
    parse_response(Response,
                   fun(Json) ->
                           { get_property_value("access_token", Json),
                             get_property_value("refresh_token", Json) }
                   end
                  ).

get_my_friends(AccessToken) ->
    Response = httpc:request(
                 get,
                 {
                   "http://api.mixi-platform.com/2/people/@me/@friends",
                   [
                    {"Authorization", lists:append("OAuth ", AccessToken)}
                   ]
                 },
                 [],
                 []),
    parse_response(Response,
                   fun(Json) ->
                           Json
                   end
                  ).

parse_response({ok, {{_, 200, _}, _, Body}}, F) ->
    {struct, Json} = mochijson2:decode(Body),
    F(Json);
parse_response({ok, {{_, Status, _}, Headers, Body}}, _) ->
    case lists:keyfind("www-authenticate", 1, Headers) of
        {_, Reason} ->
            throw(
              { Status,
                Reason }
             );
        _ ->
            {struct, Error} = mochijson2:decode(Body),
            throw(
              { Status,
                get_property_value("error", Error),
                get_property_value("error_description", Error) }
             )
    end.

get_property_value(KeyName, Json) ->
    case lists:keyfind(list_to_binary(KeyName), 1, Json) of
        {_, Value} ->
            binary_to_list(Value);
        _ ->
            []
    end.

File: Python/mixi.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from google.appengine.ext import webapp
from google.appengine.ext.webapp import util
import os
from google.appengine.ext.webapp import template
from google.appengine.ext.webapp.util import run_wsgi_app
from google.appengine.api import urlfetch
import urllib
from django.utils import simplejson

AUTHORIZATION_URL_BASE    = "https://mixi.jp/connect_authorize.pl?"
REDIRECTION_URI           = "http://localhost:8080/redirect"
TOKEN_ENDPOINT            = "https://secure.mixi-platform.com/2/token"
FRIENDS_TIMELINE_ENDPOINT = "http://api.mixi-platform.com/2/people/@me/@friends"

CONSUMER_KEY    = "*******************"
CONSUMER_SECRET = "**************************************"

class MainHandler(webapp.RequestHandler):

    def get_access_token(self,auth_code):

        token_params = {
                "grant_type":"authorization_code",
                "client_id":CONSUMER_KEY,
                "client_secret":CONSUMER_SECRET,
                "code":auth_code,
                "redirect_uri":REDIRECTION_URI,
                }
        form_data = urllib.urlencode(token_params)
        result = urlfetch.fetch(url=TOKEN_ENDPOINT,
                        payload=form_data,
                        method=urlfetch.POST,
                        headers={'Content-Type':'application/x-www-form-urlencoded'})

        json = simplejson.loads(result.content)
        return json['access_token']

    def get_friends_timeline(self,access_token):
        result = urlfetch.fetch(url=FRIENDS_TIMELINE_ENDPOINT,
                        method=urlfetch.GET,
                        headers={'Authorization':'OAuth %s'%access_token})
        return simplejson.loads(result.content)

    def get(self, mode=""):

        if mode == "login":

            query = {
                    "client_id":CONSUMER_KEY,
                    "response_type":"code",
                    "scope":"r_profile",
                    }
            url = AUTHORIZATION_URL_BASE + urllib.urlencode(query)
            self.redirect(url)

        if mode == "redirect":

            auth_code = self.request.get("code")
            access_token = self.get_access_token(auth_code)

            results = self.get_friends_timeline(access_token)

            template_values = {'results': results['entry'],}
            path = os.path.join(os.path.dirname(__file__), 'mixi.html')
            self.response.out.write(template.render(path, template_values))

def main():
    application = webapp.WSGIApplication([('/(.*)', MainHandler)],
                                         debug=True)
    util.run_wsgi_app(application)

if __name__ == '__main__':
    main()